### DISCUSSION TITLE
Blacklisted Uno R4

### DISCUSSION THREAD

**Discussion by greiman** [bxparks/AceCommon#30] [General]
I just bought a new Uno R4 and found it is "blacklisted" for AceCommon.

> c:\Users\Bill\Documents\ArduinoSdFat\libraries\AceCommon\src/AceCommon.h:43:2: error: #error Platforms using ArduinoCore-API not supported
>  #error Platforms using ArduinoCore-API not supported
>   ^~~~~
> 
> exit status 1
>
>Compilation error: exit status 1

Looks like Arduino is going all-in for the "Core-API".  What is your long term plan for blacklisting it.

The Uno R4 and many other new Arduino boards are based on Renesas MCUs.  To support this class MCU, they use the Renesas "Flexible Software Package (FSP)".  I looked at FSP and it is better than average industrial software but still huge and written by a small army of programmers.

I guess it is inevitable, Arduino could never provide all the features in FSP.  Here is the Renesas description of [FSP](https://github.com/renesas/fsp):

> FSP is the next generation Arm® MCU software package from Renesas, that enables secure devices and IoT connectivity through production ready peripheral drivers, Azure RTOS or FreeRTOS, and portable middleware stacks. FSP includes best-in-class HAL drivers with high performance and low memory footprint. Middleware stacks with Azure RTOS and FreeRTOS integration are included to ease implementation of complex modules like communication and security. The e² studio ISDE provides support with intuitive configurators and intelligent code generation to make programming and debugging easier and faster.

**Comment by bxparks** [comment]
I have no plans to support any boards using the new ArduinoCore-API in the future, unless something dramatic changes in the Arduino hardware ecosystem:

1) I don't have any interest in the Arduino hardware which use the new Arduino API (i.e. SAMD, megaAVR, UNO R4). They are usually too expensive compared to alternative boards with comparable features, and they seem far too complex (e.g. using both an ARM and ESP microcontrollers) given the functionality that they provide. I've never heard of Renesas, and I have no interest in their hardware.
2) Since I don't own any of these new Arduino boards, I cannot do any validation of my libraries on these boards. I don't want to waste my money on acquiring hardware that I am not going to use for anything.
3) As far as I can tell, the new ArduinoCore-API is undocumented, and there is no migration guide that explains how third party libraries are supposed to support both the old and new APIs. I don't have the time or patience to perform that research myself.
4) The Arduino SAMD Core migrated to the new API by going from version 1.8.9 to 1.8.10. In other words, a point release was used to migrate to a substantially different API that broke backwards compatibility. This shows that backwards compatibility is *not* a priority for Arduino. I don't want to invest my time and effort into such an unstable and unreliable ecosystem.

The only potential exception that I can see in the near future are the boards using the RP2040 processor, such as the Raspberry Pi Pico. These microcontrollers are extremely cheap and have compelling features. As far as I can tell, the most popular Arduino Core for these boards is the https://github.com/earlephilhower/arduino-pico/ core by Earle F. Philhower, which I think uses the new Arduino API. When I get some time later this year, I may look into migrating my libraries to the RP2040. The boards are so cheap, I can afford to buy a bunch of them to validate my libraries.

**Comment by greiman** [comment]
When the RP2040 came out I thought it would be my favorite Arduino board.  I bought a number of variations of the board.

I then discovered the ADC had fatal flaws for my use, less than 9 ENOB. the claim is 9.3 but careful tests get about 8.  No DAC and few other peripherals.

It's just too bare bones so I have about of a dozen RP2040 boards I am not using.

You may not like Renesas but their MCUs and MPUs are used in a huge number of products.  I hope the introduction of Renesas in the Arduino world will result in third party boards for some of the other interesting Renesas processors.

The Uno replacement boards  sells for $20.  Yes, more than the RP2040.  It is a 5V compatible replacement for AVR Uno.  Huge step up.

It has 32KB RAM, 256 KB flash. Single precision floating-point.  Four Memory Protection Units (MPUs) and a CPU stack pointer monitor function are provided for memory protection.

GPIO supports open-drain. It has up to 14-bit ADC resolution so maybe 12-bit ENOB, 12-bit DAC, four op-amps, Serial Sound Interface, CAN, RTC, Segment LCD Controller,  Capacitive Touch Sensing Unit and other stuff like AES128/256, GHASH, True Random Number Generator.

The first release supports FreeRTOS and libraries for many of the above features.

**Comment by bxparks** [comment]
Thanks for the additional info on the Renesas processors. Those are indeed interesting specs. To clarify, I don't have anything against Renesas specifically, I don't know anything about them. My main issue is with the ArduinoCore-API, which breaks backwards compatibility with the previous API, and it's not worth my time to support it.

**Comment by greiman** [comment]
> My main issue is with the ArduinoCore-API, which breaks backwards compatibility with the previous API, and it's not worth my time to support it.

It happens all the time in the world I worked in.  Huge scientific laboratories like [CERN](https://home.cern/) evolve software over decades.  API change management is formalized in the organization so APIs can change without catastrophic events.  There is a Change Management Board with representatives of system developers,  application programmers, and scientists.

This will never happen in the Arduino world.

It always amazes me how C++ evolves with many features being removed.  A new version of the standard states some feature will be available until version C++yy.  We will soon know what will change some time after 2026 as C++26 is finalized.

**Reply by bxparks** [reply]
It is astounding to me that there is a segment of the software industry where breaking API changes can be pushed out without any documentation or migration guide. But I guess the software industry is big, so anything is possible. In any case, Arduino.cc has split the Arduino API into 2 versions, dividing the associated hardware platforms into 2 sets. I'm just one guy working in a corner. I don't have the bandwidth or desire to support 2 APIs, so the new API will not be supported.

With regards to your side comment about C++, I find myself getting increasingly tired of the increasing complexity of C++. There are large parts of the language that I don't understand anymore, and when I figure something out, I cannot keep it in my head. I am constantly searching for alternatives, but in the microcontroller space, C and C++ seem like they are still king. The C language is too primitive and not expressive enough for my taste. I don't want to roll my own code for everything little thing anymore. MicroPython is too memory intensive and CPU intensive, being a garage collected interpreted language. I've looked at TinyGo, which is very interesting, but I concluded that it is not mature enough yet, and the project has very limited manpower compared to C and C++. The Zig lang is not stable, and won't be stable until 1.0 which is several years away. I plan to look at Nim and Rust. The complexity of Rust language and its slow compiler time does not give me warm feelings. Nim seems interesting on paper, but until I actually try to use it, I won't know for certain. I find myself forced to come back to C++, but I really don't like the language anymore.
